phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java

```diff
@@ -107,8 +127,37 @@ public synchronized void close() throws SQLException {
     @Override
     public synchronized Tuple next() throws SQLException {
         initScanner();
-        Tuple t = scanIterator.next();
-        return t;
+        try {
+            lastTuple = scanIterator.next();
+            if (lastTuple != null) {
+                ImmutableBytesWritable ptr = new ImmutableBytesWritable();
+                lastTuple.getKey(ptr);
+            }
+        } catch (SQLException e) {
+            try {
+                throw ServerUtil.parseServerException(e);
+            } catch(StaleRegionBoundaryCacheException e1) {
+                if(scan.getAttribute(NON_AGGREGATE_QUERY)!=null) {
+                    Scan newScan = ScanUtil.newScan(scan);
+                    if(lastTuple != null) {
+                        lastTuple.getKey(ptr);
+                        byte[] startRowSuffix = ByteUtil.copyKeyBytesIfNecessary(ptr);
+                        if(ScanUtil.isLocalIndex(newScan)) {
+                            newScan.setAttribute(SCAN_START_ROW_SUFFIX, ByteUtil.nextKey(startRowSuffix));
+                        } else {
+                            newScan.setStartRow(ByteUtil.nextKey(startRowSuffix));
+                        }
+                    }
+                    plan.getContext().getConnection().getQueryServices().clearTableRegionCache(htable.getTableName());
+                    this.scanIterator =
+                            plan.iterator(DefaultParallelScanGrouper.getInstance(), newScan);
```

Comment:
Getting the ParallelScanGrouper here as state instead of assuming the default one is being used may be necessary (or making it not necessary to have this ParallelScanGrouper - it's an edge case for the MR integration - there's likely a better way - perhaps the plan itself could know what to use for the ParallelScanGrouper?)